channelserver/src/meta.rs

```diff
@@ -0,0 +1,219 @@
+                                }), 
+                            represented_country: None, 
+                            subdivisions: Some([Subdivision { 
+                                geoname_id: Some(#), 
+                                iso_code: Some(".."), 
+                                names: Some({"lang": "name", ...}) 
+                                }]), 
+                            traits: None }
+                        }
+                    */
+                    if let Some(names) = city
+                        .city
+                        .and_then(|c: maxminddb::geoip2::model::City| c.names)
+                    {
+                        sender.city = get_preferred_language_element(&langs, names);
+                    }
+                    if let Some(names) = city
+                        .country
+                        .and_then(|c: maxminddb::geoip2::model::Country| c.names)
+                    {
+                        sender.country = get_preferred_language_element(&langs, names);
+                    }
+                    // because consistency is overrated.
+                    for subdivision in city.subdivisions {
+                        if let Some(subdivision) = subdivision.get(0) {
+                            if let Some(names) = subdivision.clone().names {
+                                sender.region = get_preferred_language_element(&langs, names);
+                                break;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        sender
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::{get_preferred_language_element, preferred_language};
+    use std::collections::BTreeMap;
+
+    #[test]
+    fn test_preferred_language() {
+        let langs = preferred_language("en-US,es;q=0.1,en;q=0.5".to_owned());
+        assert_eq!(
+            vec![
+                "en-us".to_owned(),
+                "en".to_owned(),
+                "es".to_owned(),
+                "en".to_owned(),
+            ],
+            langs
+        );
+    }
+
+    #[test]
+    fn test_get_preferred_language_element() {
+        let langs = vec![
+            "en-us".to_owned(),
+            "en".to_owned(),
+            "es".to_owned(),
+            "en".to_owned(),
+        ];
+        let bad_lang = vec!["fu".to_owned()];
+        let mut elements = BTreeMap::new();
+        elements.insert("de".to_owned(), "Kalifornien".to_owned());
+        elements.insert("en".to_owned(), "California".to_owned());
+        elements.insert("fr".to_owned(), "Californie".to_owned());
+        elements.insert("ja".to_owned(), "カリフォルニア州".to_owned());
+        assert_eq!(
+            Some("California".to_owned()),
+            get_preferred_language_element(&langs, elements.clone())
+        );
+        assert_eq!(
+            None,
+            get_preferred_language_element(&bad_lang, elements.clone())
+        );
+    }
```

Comment:
How much effort would it be to add a test for the ip-lookup stuff here?  IIUC that can get a bit gnarly in rust because it's hard to mock out the concrete implementation with one that produces known good test results, but it's a shame to leave that logic untested...

Reply:
It's possible, since we can just hand it a bunch of known addresses, although it seems like this might be "testing static". If we don't trust the maxminddb API layer, I can understand adding the tests, but we might want to do the same for things like the Actix header libraries and what-not, and that could get ugly fast.

Reply:
I trust the maxmind layer, I was mostly thinking about our own branching logic here.